modules/priceFloors.js

```diff
@@ -0,0 +1,643 @@
+    // storing exact matches as lowerCase since we want to compare case insensitively
+    accum.push(exactMatch === '*' ? ['*'] : [exactMatch.toLowerCase(), '*']);
+    return accum;
+  }, []);
+}
+
+/**
+ * @summary get's the first matching floor based on context provided.
+ * Generates all possible rule matches and picks the first matching one.
+ */
+export function getFirstMatchingFloor(floorData, bidObject, responseObject = {}) {
+  let fieldValues = enumeratePossibleFieldValues(utils.deepAccess(floorData, 'schema.fields') || [], bidObject, responseObject);
+  if (!fieldValues.length) return { matchingFloor: floorData.default };
+
+  // look to see iof a request for this context was made already
+  let matchingInput = fieldValues.map(field => field[0]).join('-');
+  // if we already have gotten the matching rule from this matching input then use it! No need to look again
+  let previousMatch = utils.deepAccess(floorData, `matchingInputs.${matchingInput}`);
+  if (previousMatch) {
+    return previousMatch;
+  }
+  let allPossibleMatches = generatePossibleEnumerations(fieldValues, utils.deepAccess(floorData, 'schema.delimiter') || '|');
+  let matchingRule = find(allPossibleMatches, hashValue => floorData.values.hasOwnProperty(hashValue));
+
+  let matchingData = {
+    matchingFloor: floorData.values[matchingRule] || floorData.default,
+    matchingData: allPossibleMatches[0], // the first possible match is an "exact" so contains all data relevant for anlaytics adapters
+    matchingRule
+  };
+  // save for later lookup if needed
+  utils.deepSetValue(floorData, `matchingInputs.${matchingInput}`, {...matchingData});
+  return matchingData;
+}
+
+/**
+ * @summary Generates all possible rule hash's based on input array of array's
+ * The generated list is of all possible key matches based on fields input
+ * The list is sorted by least amount of * in rule to most with left most fields taking precedence
+ */
+function generatePossibleEnumerations(arrayOfFields, delimiter) {
+  return arrayOfFields.reduce((accum, currentVal) => {
+    let ret = [];
+    accum.map(obj => {
+      currentVal.map(obj1 => {
+        ret.push(obj + delimiter + obj1)
+      });
+    });
+    return ret;
+  }).sort((left, right) => left.split('*').length - right.split('*').length);
+}
+
+/**
+ * @summary If a the input bidder has a registered cpmadjustment it returns the input CPM after being adjusted
+ */
+export function getBiddersCpmAdjustment(bidderName, inputCpm) {
+  const adjustmentFunction = utils.deepAccess(getGlobal(), `bidderSettings.${bidderName}.bidCpmAdjustment`);
+  if (adjustmentFunction) {
+    return parseFloat(adjustmentFunction(inputCpm));
+  }
+  return parseFloat(inputCpm);
+}
+
+/**
+ * @summary This function takes the original floor and the adjusted floor in order to determine the bidders actual floor
+ */
+export function calculateAdjustedFloor(oldFloor, newFloor) {
+  return oldFloor / newFloor * oldFloor;
```

Comment:
Yeah thanks for pointing it out.

Is the recommendation to do something similar to what the cpmBucket function is doing?

The input to this function will more often than not be the same two values, The only time it will have differing values is if the bidder in question has a cpmAdjustment function set in bidderSettings.

Reply:
In terms of the recommendation, I think what we do in cpmBucket (ie use big numbers for division) was the way to go.

I think it still may worth updating just to avoid future edge cases, but anyone else please chime in here if it doesn't seem like a big deal.